pkg/pillar/hypervisor/kvm.go

```diff
@@ -462,20 +462,135 @@ func (ctx kvmContext) Task(status *types.DomainStatus) types.Task {
 	}
 }
 
-// TBD: Have a better way to calculate this number.
-// For now it is based on some trial-and-error experiments.
-// Container limit is reduced to 100MiB.
-func minVMMOverhead(config types.DomainConfig) int64 {
+func estimatedVMMOverhead(domainName string, config types.DomainConfig, aa *types.AssignableAdapters) (int64, error) {
+	// Container limit is set to 100MiB.
+	// This is a temporary solution until we have a better way to predict
+	// the memory usage of the container.
 	if config.IsOCIContainer() {
-		return 100 << 20 // Mb in bytes
+		return 100 << 20, nil // Mb in bytes
 	}
-	return 600 << 20 // Mb in bytes
-}
 
-func vmmOverhead(config types.DomainConfig,
-	globalConfig *types.ConfigItemValueMap) (int64, error) {
 	var overhead int64
 
+	mmioOverhead, err := mmioVMMOverhead(domainName, config, aa)
+
+	if err != nil {
+		return 0, logError("mmioVMMOverhead() failed for domain %s: %v",
+			domainName, err)
+	} else {
+		overhead = undefinedVMMOverhead(config) + ramVMMOverhead(config) +
+			qemuVMMOverhead(config) + cpuVMMOverhead(config) + mmioOverhead
+	}
+
+	return overhead, nil
+}
+
+func ramVMMOverhead(config types.DomainConfig) int64 {
+	// 2.5% of the total RAM alocated for VM in bytes
+	// this formula is precise and well explained in the following article:
+	// TBD
+	return int64(config.Memory) * 1024 * 25 / 1000
+}
+
+// overhead for qemu binaries and libraries
+func qemuVMMOverhead(config types.DomainConfig) int64 {
+	return 20 << 20 // Mb in bytes
+}
+
+// overhead for VMM memory mapped IO
+// it fluctuates between 0.66 and 0.81 % of MMIO total size
+// for all mapped devices. Set it to 1% to be on the safe side
+// this can be a pretty big number for GPUs with very big
+// aperture size (e.g. 64G for NVIDIA A40)
+func mmioVMMOverhead(domainName string, config types.DomainConfig, aa *types.AssignableAdapters) (int64, error) {
+	var pciAssignments []typeAndPCI
+	var mmioSize uint64 = 0
+
+	for _, adapter := range config.IoAdapterList {
+		logrus.Debugf("processing adapter %d %s\n", adapter.Type, adapter.Name)
+		list := aa.LookupIoBundleAny(adapter.Name)
+		// We reserved it in handleCreate so nobody could have stolen it
+		if len(list) == 0 {
+			return 0, logError("IoBundle disappeared %d %s for %s\n",
+				adapter.Type, adapter.Name, domainName)
+		}
+		for _, ib := range list {
+			if ib == nil {
+				continue
+			}
+			if ib.UsedByUUID != config.UUIDandVersion.UUID {
+				return 0, logError("IoBundle not ours %s: %d %s for %s\n",
+					ib.UsedByUUID, adapter.Type, adapter.Name,
+					domainName)
+			}
+			if ib.PciLong != "" {
+				logrus.Infof("Adding PCI device <%v>\n", ib.PciLong)
```

Comment:
```suggestion
				logrus.Infof("Adding PCI device <%s>\n", ib.PciLong)
```
isn't it just a string?

Reply:
yep. will fix